scene/resources/material.cpp

```diff
@@ -954,132 +1005,213 @@ void BaseMaterial3D::_update_shader() {
 		code += "varying vec3 uv2_triplanar_pos;\n";
 	}
 	if (flags[FLAG_UV1_USE_TRIPLANAR]) {
-		code += "uniform float uv1_blend_sharpness;\n";
-		code += "varying vec3 uv1_power_normal;\n";
+		code += R"(
+uniform float uv1_blend_sharpness : hint_range(0.0, 150.0, 0.001);
+varying vec3 uv1_power_normal;
+)";
 	}
 
 	if (flags[FLAG_UV2_USE_TRIPLANAR]) {
-		code += "uniform float uv2_blend_sharpness;\n";
-		code += "varying vec3 uv2_power_normal;\n";
+		code += R"(uniform float uv2_blend_sharpness : hint_range(0.0, 150.0, 0.001);
+varying vec3 uv2_power_normal;
+)";
 	}
 
-	code += "uniform vec3 uv1_scale;\n";
-	code += "uniform vec3 uv1_offset;\n";
-	code += "uniform vec3 uv2_scale;\n";
-	code += "uniform vec3 uv2_offset;\n";
+	code += R"(
+uniform vec3 uv1_scale;
+uniform vec3 uv1_offset;
+uniform vec3 uv2_scale;
+uniform vec3 uv2_offset;
+)";
 
-	code += "\n\n";
-
-	code += "void vertex() {\n";
+	// Generate vertex shader.
+	code += R"(
+void vertex() {
+)";
 
 	if (flags[FLAG_SRGB_VERTEX_COLOR]) {
-		code += "	if (!OUTPUT_IS_SRGB) {\n";
-		code += "		COLOR.rgb = mix(pow((COLOR.rgb + vec3(0.055)) * (1.0 / (1.0 + 0.055)), vec3(2.4)), COLOR.rgb * (1.0 / 12.92), lessThan(COLOR.rgb, vec3(0.04045)));\n";
-		code += "	}\n";
+		code += R"(	// Vertex Color is sRGB: Enabled
```

Comment:
I think there should be a line break at the start here, and none in the previous R-string after `vertex() {`.

Remember that anything which is after an `if` is conditional, so this may not be part of a shader, and the next potential chunks of code all start with an extra line break, which will lead to e.g.:
```
vertex() {

	// Use Point Size: Enabled
	POINT_SIZE = point_size;
```